lib/posix/eventfd.c

```diff
@@ -223,20 +261,29 @@ int eventfd(unsigned int initval, int flags)
 	}
 
 	if (efd == NULL) {
-		z_free_fd(fd);
 		errno = ENOMEM;
-		k_mutex_unlock(&eventfd_mtx);
-		return -1;
+		goto exit_mtx;
+	}
+
+	fd = z_reserve_fd();
+	if (fd < 0) {
+		goto exit_mtx;
 	}
 
 	efd->flags = EFD_IN_USE | flags;
-	efd->cnt = 0;
-	k_sem_init(&efd->read_sem, 0, 1);
-	k_sem_init(&efd->write_sem, 0, UINT32_MAX);
+	efd->cnt = initval;
+
+	k_poll_signal_init(&efd->write_sig);
+	k_poll_signal_init(&efd->read_sig);
+	z_waitq_init(&efd->wait_q);
+
+	if (initval != 0)
+		k_poll_signal_raise(&efd->read_sig, 0);
+	k_poll_signal_raise(&efd->write_sig, 0);
```

Comment:
if (initval < UINT64_MAX - 1)

Reply:
Initval is uint32_t, so we don't have to check it. It's some oddity.